core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,250 @@
+        val usedMem = maxMem - remainingMem
+        val toBeReplicatedMem = usedMem - alreadyReplicated
+        execId -> toBeReplicatedMem
+      }.toSeq.sortBy { case (k, v) => v }
+
+    bytesToReplicate.scan(("start", freeMemOnRemaining)) {
+      case ((_, remainingMem), (execId, mem)) => (execId, remainingMem - mem)
+    }.drop(1)
+      .toMap
+      .mapValues(freeMem => freeMem >= 0)
+  }
+
+  /**
+   * Gets a list of cached blocks from the block manager master and updates.
+   * CacheRecoveryManagerState.
+   *
+   * @param execId an executor Id
+   */
+  private def updateBlockState(execId: String): Unit = {
+    val blocks: mutable.Set[RDDBlockId] = blockManagerMasterEndpoint
+      .askSync[collection.Set[BlockId]](GetCachedBlocks(execId))
+      .flatMap(_.asRDDId)(collection.breakOut)
+
+    blocks --= savedBlocks.getOrElse(execId, new mutable.HashSet)
+    blocksToSave(execId) = mutable.PriorityQueue[RDDBlockId](blocks.toSeq: _*)(Ordering.by(_.rddId))
+    killTimers.getOrElseUpdate(execId, scheduleKill(new KillRunner(execId)))
+  }
+
+  class KillRunner(execId: String) extends Runnable { def run(): Unit = killExecutor(execId) }
+  private def scheduleKill(k: KillRunner) = scheduler.schedule(k, forceKillAfterS, TimeUnit.SECONDS)
+
+  /**
+   * Ask block manager master to replicate one cached block.
+   *
+   * @param execId the executor to replicate a block from
+   * @return false if there are no more blocks or if there is an issue
+   */
+  def replicateFirstBlock(execId: String): (Future[Boolean], Option[RDDBlockId]) = synchronized {
+    logDebug(s"Replicate block on executor $execId.")
+    blocksToSave.get(execId) match {
+      case Some(p) if p.nonEmpty => doReplication(execId, p.dequeue())
+      case _ => (Future.successful(false), None)
+    }
+  }
+
+  private def doReplication(
+      execId: String,
+      blockId: RDDBlockId): (Future[Boolean], Option[RDDBlockId]) = {
+    val savedBlocksForExecutor = savedBlocks.getOrElseUpdate(execId, new mutable.HashSet)
+    savedBlocksForExecutor += blockId
+    val replicateMessage = ReplicateOneBlock(execId, blockId, blocksToSave.keys.toSeq)
+    logTrace(s"Started replicating block $blockId on exec $execId.")
+    val future = blockManagerMasterEndpoint.ask[Boolean](replicateMessage)
+    (future, Some(blockId))
+  }
+
+  /**
+   * Ask ExecutorAllocationManager to kill executor and clean up state
+   * Note executorAllocationManger.killExecutors is blocking, this function should be fixed
+   * to use a non blocking version
+   * @param execId the executor to kill
+   */
+  def killExecutor(execId: String): Unit = synchronized {
+    logDebug(s"Send request to kill executor $execId.")
+    killTimers.remove(execId).foreach(_.cancel(false))
+    blocksToSave -= execId
+    savedBlocks -= execId
+    executorAllocationManager.killExecutors(Seq(execId), forceIfPending = true)
+  }
+
+  def stop(): java.util.List[Runnable] = scheduler.shutdownNow()
```

Comment:
Same thing, don't see you using the return value anywhere, so `Unit`.

Reply:
fixed